cardano-node/src/Cardano/Node/Shutdown.hs

```diff
@@ -0,0 +1,191 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Cardano.Node.Shutdown
+  (
+  -- * Generalised shutdown handling
+    ShutdownFDs
+  , withShutdownHandling
+
+  -- * Requesting shutdown 
+  , ShutdownDoorbell
+  , getShutdownDoorbell
+  , triggerShutdown
+
+  -- * Handle requests to shutdown on ChainDB slot sync limit.
+  , maybeSpawnOnSlotSyncedShutdownHandler
+  )
+where
+
+import           Cardano.Prelude hiding (ByteString, atomically, take, trace)
+
+import qualified Control.Concurrent.Async as Async
+import           Data.Text (pack)
+
+import qualified GHC.IO.Handle.FD as IO (fdToHandle)
+import qualified System.Process as IO (createPipeFd)
+import qualified System.IO as IO
+import qualified System.IO.Error  as IO
+
+import           Cardano.BM.Trace
+import           Cardano.BM.Data.Tracer (
+                     TracingVerbosity (..), TracingFormatting (..),
+                     severityNotice, trTransformer)
+import           Cardano.Slotting.Slot (WithOrigin(..))
+import           Control.Tracer
+import qualified Ouroboros.Consensus.Storage.ChainDB as ChainDB
+import           Ouroboros.Consensus.Util.ResourceRegistry (ResourceRegistry)
+import           Ouroboros.Consensus.Util.STM (onEachChange)
+import           Ouroboros.Network.Block (MaxSlotNo(..), SlotNo, pointSlot)
+
+import           Cardano.Config.Types
+
+-- | 'ShutdownFDs' mediate the graceful shutdown requests,
+-- either external or internal to the process.
+-- 
+-- In the external mediation case, the parent process passes us the file descriptor
+-- number of the read end of a pipe, via the CLI with @--shutdown-ipc FD@.
+-- In the internal mediation case, we create our own pipe.
+-- In both cases we store the accessible ends in 'ShutdownFDs'.
+--
+-- In either case, if the write end gets closed, either deliberately:
+--   - by internal call of 'triggerShutdown' on 'ShutdownFDs', or
+--   - by parent process
+-- ..or automatically, because the parent process itself terminated,
+-- then we initiate a clean shutdown.
+data ShutdownFDs
+  = NoShutdownFDs
+  | ExternalShutdown !ShutdownListener
+  -- ^ Extra-processually signalled shutdown.
+  | InternalShutdown !ShutdownListener !ShutdownDoorbell
+  -- ^ Intra-processually signalled shutdown.
```

Comment:
Not sure about the name -- it reflects content, not intent.

Maybe `ShutdownChan` or just `Shutdown` instead?